pkg/api/parser/protobuf/protobuf.go

```diff
@@ -12,14 +13,24 @@ import (
 // ParseRequest is responsible for populating the write request from the
 // data in the request in protobuf format.
 func ParseRequest(r *http.Request, wr *prompb.WriteRequest) error {
-	reqBuf, err := ioutil.ReadAll(r.Body)
+	b := bufPool.Get().(*bytes.Buffer)
+	b.Reset()
+
+	_, err := b.ReadFrom(r.Body)
```

Comment:
Can we do the same for read requests? That will help us save extra allocations.

Reply:
Yes but in a separate PR. We should get flamegraphs to see we need this before doing it

Reply:
@cevian TODO.